runtimes/pythonrt/ak_runner/deterministic.py

```diff
@@ -0,0 +1,113 @@
+"""List of functions/modules that are deterministic should not run as activities."""
+
+# Function/modules added here should be:
+# - Stateless
+# - Common enough to be included
+# So, `re` is OK, `random` is not
+
+import datetime
+
+
+def is_determinstic(fn):
+    """Return True if fn (callable) can run outside of activity."""
+    return fn in functions or fn.__module__ in modules
+
+
+# Modules are represented as strings func.__module__ is a string
+modules = {
```

Comment:
Allow-list vs. deny-list question: would it be simpler to assume that most builtins are deterministic and exclude known/suspected exceptions, or not? I'm asking becaus "Common enough to be included" seems too subjective/controversial.

i.e.

```
stdlib_modules = frozenset({m for m in sys.stdlib_module_names if not m.startswith("_")})
modules = stdlib_modules - frozenset({
    ...
})
```

For reference, here are the stdlib modules not mentioned in this list:

aifc
antigravity
argparse
ast
asyncio
atexit
audioop
bdb
binascii
cProfile
calendar
cgi
cgitb
chunk
cmd
code
codecs
codeop
colorsys
compileall
concurrent
configparser
contextvars
copyreg
crypt
ctypes
curses
datetime
dbm
difflib
dis
doctest
email
encodings
ensurepip
errno
faulthandler
fcntl
filecmp
fileinput
ftplib
gc
genericpath
getopt
getpass
gettext
glob
grp
hmac
http
idlelib
imaplib
imghdr
importlib
inspect
io
keyword
lib2to3
linecache
locale
logging
mailbox
mailcap
marshal
mimetypes
mmap
modulefinder
msilib
msvcrt
multiprocessing
netrc
nis
nntplib
nt
ntpath
nturl2path
numbers
opcode
optparse
os
ossaudiodev
pathlib
pdb
pickle
pickletools
pipes
pkgutil
platform
plistlib
poplib
posix
posixpath
profile
pstats
pty
pwd
py_compile
pyclbr
pydoc
pydoc_data
pyexpat
queue
quopri
random
readline
reprlib
resource
rlcompleter
runpy
sched
secrets
select
selectors
shelve
shutil
signal
site
smtplib
sndhdr
socket
socketserver
spwd
sqlite3
sre_compile
sre_constants
sre_parse
ssl
stat
string
stringprep
subprocess
sunau
symtable
sys
sysconfig
syslog
tabnanny
tarfile
telnetlib
tempfile
termios
textwrap
this
threading
time
timeit
tkinter
token
tokenize
trace
tracemalloc
tty
turtle
turtledemo
types
typing
unittest
urllib
uu
uuid
venv
warnings
wave
weakref
webbrowser
winreg
winsound
wsgiref
xdrlib
xml
xmlrpc
zipapp
zipfile
zipimport
zoneinfo

Reply:
I've added types, typing,  xml, zipfile and zoneinfo.

Others:
- unittest is problematic (especially the main function). And IMO most developers don't use code from unittest in their code - what do you want from there?
- uuid is not determinstic
- I've added zoneinfo, but you might get different results if the zone db changes. IMO it's OK

Reply:
I already approved the PR, because this list is a suggestion and not a strong requirement, so this is fine, but just to respond:

* `unittest` - I want people to have a clear and familiar way to unit-test their code, so as long as it works that's fine; I see [you already use it](https://github.com/autokitteh/samples/pull/84/files#diff-cc0fd32bb3d003d46c443c67953575c8d1124293552175b91ef784791fc77dce), but I'd like to know more about the user experience, anyway not a high priority
* UUID - I see, I agree with you; I'll check if we need to handle it with function granularity or not, like datetime
* `zoneinfo` - ack, I agree